src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java

```diff
@@ -512,17 +479,114 @@ public Gson getGson() {
     /**
      * Executes a HTTP request and return a domain object
      *
-     * @param request  The HTTP request to execute.
-     * @param Class<T> class of domain object to return
-     * @return <T> T
+     * @param connection  The HTTP connection to execute.
+     * @return <T> T json object
      */
-    public <T> T executeRequest(HttpRequestBase request, Class<T> classType) {
-        HttpResponse response = null;
+    public InputStream executeRequest(HttpConnection connection) {
+        return executeToInputStream(connection);
+    }
+
+    // - if 2xx then return stream
+    // - map 404 to NoResourceException
+    // - if there's a couch error returned as json, un-marshall and throw
+    // - anything else, just throw the IOException back, use the cause part of the exception?
+
+    // it needs to catch eg FileNotFoundException and rethrow to emulate the previous exception handling behaviour
+    private InputStream executeToInputStream(HttpConnection connection) throws CouchDbException {

```

Comment:
I think there would be some value to extracting this into a class of its own and potentially with a response object (of a new type) instead of `InputStream`, but I think that can happen later.